### PR TITLE
electra engine api support

### PR DESCRIPTION
**Pull request by kasey**
**What type of PR is this?**

Feature

**What does this PR do? Why is it needed?**

Adds support for interacting with the prague engine api to support the electra fork.

### CODE REVIEW

File: beacon-chain/execution/engine_client.go [hunk 1]
```diff
@@ -828,8 +790,8 @@ func fullPayloadFromPayloadBody(
 		if err != nil {
 			return nil, errors.Wrap(err, "unable to extract BlobGasUsed attribute from execution payload header")
 		}
-		return blocks.WrappedExecutionPayloadDeneb(
-			&pb.ExecutionPayloadDeneb{
+		return blocks.WrappedExecutionPayloadElectra(
+			&pb.ExecutionPayloadElectra{
```
Comment (terencechain) [inline beacon-chain/execution/engine_client.go:794]: We are missing `DepositReceipts` and `WithdrawalRequests` here. I think we will need an `ExecutionPayloadBodyV2` and a new version of `engine_getPayloadBodiesByRoot` to accompany that

File: proto/engine/v1/json_marshal_unmarshal.go [hunk 2]
```diff
@@ -746,6 +896,224 @@ func (e *ExecutionPayloadDeneb) MarshalJSON() ([]byte, error) {
...
+		Withdrawals:        withdrawals,
+		BlobGasUsed:        &blobGasUsed,
+		ExcessBlobGas:      &excessBlobGas,
+		WithdrawalRequests: WithdrawalRequestProtoToJson(e.WithdrawalRequests),
+		DepositRequests:    DepositRequestProtoToJson(e.DepositReceipts),
+	})
+}
+
+func (j *ExecutionPayloadElectraJSON) ElectraDepositReceipts() []*DepositReceipt {
+	rcpt := make([]*DepositReceipt, 0, len(j.DepositRequests))
```
Comment (terencechain) [inline proto/engine/v1/json_marshal_unmarshal.go:948]: Is there a reason why we dont want to define the size `make([]*DepositReceipt, len(j.DepositRequests))`

Comment (kasey) [reply proto/engine/v1/json_marshal_unmarshal.go:948]: Not really, I can write it the other way if you prefer!

Comment (kasey) [reply proto/engine/v1/json_marshal_unmarshal.go]: changed it create the fixed size slices and assign to index

File: beacon-chain/execution/engine_client.go (hunk 1 above)
Comment (kasey) [reply beacon-chain/execution/engine_client.go:794]: This is going to be a bigger change and still a little ambiguous with upstream specs. Can we handle this in a follow-up PR?

File: proto/engine/v1/json_marshal_unmarshal.go [hunk 3]
```diff
@@ -254,6 +295,94 @@ type GetPayloadV3ResponseJson struct {
...
+	if j.GasLimit == nil {
+		return errors.New("missing required field 'gasLimit' for ExecutionPayload")
+	}
+	if j.BlobGasUsed == nil {
+		return errors.New("missing required field 'blobGasUsed' for ExecutionPayload")
+	}
+	if j.ExcessBlobGas == nil {
+		return errors.New("missing required field 'excessBlobGas' for ExecutionPayload")
+	}
+	return nil
```
Comment (terencechain) [inline proto/engine/v1/json_marshal_unmarshal.go:383]: Do we need check for nils here for deposit receipts and withdrawal requests?

Comment (kasey) [reply proto/engine/v1/json_marshal_unmarshal.go:383]: I think if those fields are not present in the EL response, their keys could be omitted in the response? In which case the unmarshaled fields would be an empty list or possibly nil.

If those fields are either nil or empty lists, eg `ElectraExecutionLayerWithdrawalRequests` will return an empty list as `len` and `range` are safe on nil list values. So I think it's safe for them to be nil as far as the unmarshaling process is concerned, and I think we probably don't want to treat such a payload as invalid. This is equivalent to how we deal with Withdrawals:
```
if dec.ExecutionPayload.Withdrawals == nil {
	dec.ExecutionPayload.Withdrawals = make([]*Withdrawal, 0)
}
```

**Review by terencechain — APPROVED** [review]
Lgtm

**Pull request merged**